File: common/src/main/java/org/sqlbuilder/common/Resolver2.java
package org.sqlbuilder.common;

import java.util.Map;
import java.util.function.BiFunction;

public abstract class Resolver2<T, U, R> implements BiFunction<T, U, R>
{
	public final Map<T, Map<U, R>> map;

	public Resolver2(final Map<T, Map<U, R>> map)
	{
		this.map = map;
	}

	@Nullable
	@Override
	public R apply(final T k, U k2)
	{
		var m2 = map.get(k);
		if (m2 == null)
		{
			return null;
		}
		return m2.get(k2);
	}
}

File: common/src/main/java/org/sqlbuilder/common/Resolver2.kt
package org.sqlbuilder.common

abstract class Resolver2<T, U, R>(val map: Map<T, Map<U, R>>) : (T, U) -> R? {

    @Nullable
    override fun invoke(k: T, k2: U): R? {
        return map[k]?.get(k2)
    }
}
